File: src/priv/bajdcc/LALR1/grammar/codegen/Codegen.java
package priv.bajdcc.LALR1.grammar.codegen;

import priv.bajdcc.LALR1.grammar.runtime.*;
import priv.bajdcc.LALR1.grammar.symbol.SymbolTable;
import priv.bajdcc.LALR1.grammar.tree.Function;
import priv.bajdcc.util.HashListMap;
import priv.bajdcc.util.HashListMapEx;
import priv.bajdcc.util.intervalTree.Interval;

import java.util.ArrayList;
import java.util.List;

/**
 * 【中间代码】中间代码接口实现
 *
 * @author bajdcc
 */
public class Codegen implements ICodegen, ICodegenBlock, ICodegenByteWriter {

	private HashListMap<Object> symbolList;
	private HashListMap<Function> funcMap = new HashListMap<>();
	private CodegenData data = new CodegenData();
	private RuntimeDebugInfo info = new RuntimeDebugInfo();
	private List<Byte> insts = null;
	private List<Interval<Object>> itvList = new ArrayList<>();

	public Codegen(SymbolTable symbol) {
		symbolList = symbol.getManageDataService().getSymbolList();
		for (ArrayList<Function> funcs : symbol.getManageDataService().getFuncMap().list) {
			for (Function func : funcs) {
				funcMap.add(func);
			}
		}
	}

	@Override
	public int getFuncIndex(Function func) {
		return funcMap.indexOf(func);
	}

	/**
	 * 产生中间代码
	 */
	public void gencode() {
		genCode(RuntimeInst.iopena);
		genCodeWithFuncWriteBack(RuntimeInst.ipush, 0);
		genCode(RuntimeInst.icall);
		genCode(RuntimeInst.ipop);
		genCode(RuntimeInst.ihalt);
		genCode(RuntimeInst.inop);
		for (Function func : funcMap.list) {
			if (func.isExtern()) {
				info.addExports(func.getRealName(), data.getCodeIndex());
				info.addExternalFunc(func.getRealName(), new CodegenFuncDoc(
						func.getDoc()));
			}
			func.genCode(this);
		}
	}

	/**
	 * 产生代码页
	 * 
	 * @return 代码页
	 */
	public RuntimeCodePage genCodePage() {
		List<Object> objs = new ArrayList<>(symbolList.list);
		insts = new ArrayList<>();
		for (RuntimeInstUnary unary : data.callsToWriteBack) {
			unary.op1 = data.funcEntriesMap.get(funcMap.list.get(unary.op1)
                    .getRefName());
        }
		for (RuntimeInstBase inst : data.insts) {
			inst.gen(this);
		}
		return new RuntimeCodePage(objs, insts, info, itvList);
	}

	@Override
	public void genFuncEntry(String funcName) {
		data.pushFuncEntry(funcName);
		info.addFunc(funcName, data.getCodeIndex());
	}

	@Override
	public RuntimeInstNon genCode(RuntimeInst inst) {
		RuntimeInstNon in = new RuntimeInstNon(inst);
		data.pushCode(in);
		return in;
	}

	@Override
	public RuntimeInstUnary genCode(RuntimeInst inst, int op1) {
		RuntimeInstUnary in = new RuntimeInstUnary(inst, op1);
		data.pushCode(in);
		return in;
	}

	@Override
	public RuntimeInstUnary genCodeWithFuncWriteBack(RuntimeInst inst, int op1) {
		RuntimeInstUnary in = new RuntimeInstUnary(inst, op1);
		data.pushCodeWithFuncWriteBack(in);
		return in;
	}

	@Override
	public RuntimeInstBinary genCode(RuntimeInst inst, int op1, int op2) {
		RuntimeInstBinary in = new RuntimeInstBinary(inst, op1, op2);
		data.pushCode(in);
		return in;
	}

	@Override
	public ICodegenBlock getBlockService() {
		return this;
	}

	@Override
	public void genDebugInfo(int start, int end, Object info) {
		itvList.add(new Interval<Object>(start, end, info));
	}

	@Override
	public RuntimeInstUnary genBreak() {
		CodegenBlock block = data.getBlock();
		if (block == null) {
			return null;
		}
		RuntimeInstUnary in = new RuntimeInstUnary(RuntimeInst.ijmp,
				block.breakId);
		data.pushCode(in);
		return in;
	}

	@Override
	public RuntimeInstUnary genContinue() {
		CodegenBlock block = data.getBlock();
		if (block == null) {
			return null;
		}
		RuntimeInstUnary in = new RuntimeInstUnary(RuntimeInst.ijmp,
				block.continueId);
		data.pushCode(in);
		return in;
	}

	@Override
	public void enterBlockEntry(CodegenBlock block) {
		data.enterBlockEntry(block);
	}

	@Override
	public void leaveBlockEntry() {
		data.leaveBlockEntry();
	}

	@Override
	public boolean isInBlock() {
		return data.hasBlock();
	}

	@Override
	public int genDataRef(Object object) {
		return symbolList.put(object);
	}

	@Override
	public int getCodeIndex() {
		return data.getCodeIndex();
	}

	public String getCodeString() {
		StringBuilder sb = new StringBuilder();
		sb.append("#### 中间代码 ####");
		sb.append(System.lineSeparator());
		int idx = 0;
		for (RuntimeInstBase inst : data.insts) {
			sb.append(String.format("%03d\t%s", idx, inst.toString("\t")));
			idx += inst.getAdvanceLength();
			sb.append(System.lineSeparator());
		}
		return sb.toString();
	}

	@Override
	public String toString() {
		return getCodeString();
	}

	@Override
	public void genInst(RuntimeInst inst) {
		insts.add((byte) inst.ordinal());
	}

	@Override
	public void genOp(int op) {
		for (int i = 0; i < 4; i++) {
			insts.add((byte) (op >> 8 * i & 0xFF));
		}
	}
}

File: src/main/java/com/bajdcc/LALR1/grammar/codegen/Codegen.kt
package com.bajdcc.LALR1.grammar.codegen

import com.bajdcc.LALR1.grammar.runtime.*
import com.bajdcc.LALR1.grammar.symbol.SymbolTable
import com.bajdcc.LALR1.grammar.tree.Func
import com.bajdcc.util.HashListMap
import com.bajdcc.util.intervalTree.Interval

/**
 * 【中间代码】中间代码接口实现
 *
 * @author bajdcc
 */
class Codegen(symbol: SymbolTable) : ICodegen, ICodegenBlock, ICodegenByteWriter {

    private val symbolList = symbol.manageDataService.symbolList
    private val funcMap = HashListMap<Func>()
    private val data = CodegenData()
    private val info = RuntimeDebugInfo()
    private val insts = mutableListOf<Byte>()
    private val itvList = mutableListOf<Interval<Any>>()

    val codeString: String
        get() {
            val sb = StringBuilder()
            sb.append("#### 中间代码 ####")
            sb.append(System.lineSeparator())
            var idx = 0
            data.insts.forEach {
                sb.append(String.format("%03d\t%s", idx, it.toString("\t")))
                idx += it.advanceLength
                sb.append(System.lineSeparator())
            }
            return sb.toString()
        }

    init {
        for (funcs in symbol.manageDataService.funcMap.toList()) {
            for (func in funcs) {
                funcMap.add(func)
            }
        }
    }

    override fun getFuncIndex(func: Func): Int {
        return funcMap.indexOf(func)
    }

    /**
     * 产生中间代码
     */
    fun gencode() {
        genCode(RuntimeInst.iopena)
        genCodeWithFuncWriteBack(RuntimeInst.ipush, 0)
        genCode(RuntimeInst.icall)
        genCode(RuntimeInst.ipop)
        genCode(RuntimeInst.ihalt)
        genCode(RuntimeInst.inop)
        funcMap.forEach { func ->
            if (func.isExtern) {
                info.addExports(func.realName, data.codeIndex)
                info.addExternalFunc(func.realName,
                        RuntimeDebugExec(paramsDoc = func.params.joinToString(separator = "，", transform = { it.toRealString() })))
            }
            func.genCode(this)
        }
    }

    /**
     * 产生代码页
     *
     * @return 代码页
     */
    fun genCodePage(): RuntimeCodePage {
        val objs = symbolList.toList()
        data.callsToWriteBack.forEach { unary ->
            unary.op1 = data.funcEntriesMap[funcMap.get(unary.op1).refName]!!
        }
        for (inst in data.insts) {
            inst.gen(this)
        }
        return RuntimeCodePage(objs, insts.toByteArray(), info, itvList)
    }

    override fun genFuncEntry(funcName: String) {
        data.pushFuncEntry(funcName)
        info.addFunc(funcName, data.codeIndex)
    }

    override fun genCode(inst: RuntimeInst): RuntimeInstNon {
        val i = RuntimeInstNon(inst)
        data.pushCode(i)
        return i
    }

    override fun genCode(inst: RuntimeInst, op1: Int): RuntimeInstUnary {
        val i = RuntimeInstUnary(inst, op1)
        data.pushCode(i)
        return i
    }

    override fun genCodeWithFuncWriteBack(inst: RuntimeInst, op1: Int): RuntimeInstUnary {
        val i = RuntimeInstUnary(inst, op1)
        data.pushCodeWithFuncWriteBack(i)
        return i
    }

    override fun genCode(inst: RuntimeInst, op1: Int, op2: Int): RuntimeInstBinary {
        val i = RuntimeInstBinary(inst, op1, op2)
        data.pushCode(i)
        return i
    }

    override fun genDebugInfo(start: Int, end: Int, info: Any) {
        itvList.add(Interval((start - 1).toLong(), (end + 1).toLong(), info))
    }

    override val blockService: ICodegenBlock
        get() {
            return this
        }

    override fun genBreak(): RuntimeInstUnary {
        val block = data.block
        val i = RuntimeInstUnary(RuntimeInst.ijmp,
                block.breakId)
        data.pushCode(i)
        return i
    }

    override fun genContinue(): RuntimeInstUnary {
        val block = data.block
        val i = RuntimeInstUnary(RuntimeInst.ijmp,
                block.continueId)
        data.pushCode(i)
        return i
    }

    override fun enterBlockEntry(block: CodegenBlock) {
        data.enterBlockEntry(block)
    }

    override fun leaveBlockEntry() {
        data.leaveBlockEntry()
    }

    override val isInBlock: Boolean
        get() {
            return data.hasBlock()
        }

    override fun genDataRef(obj: Any): Int {
        return symbolList.put(obj)
    }

    override val codeIndex: Int
        get() {
            return data.codeIndex
        }

    override fun toString(): String {
        return codeString
    }

    override fun genInst(inst: RuntimeInst) {
        insts.add(inst.ordinal.toByte())
    }

    override fun genOp(op: Int) {
        for (i in 0..3) {
            insts.add((op shr 8 * i and 0xFF).toByte())
        }
    }
}
